File: ios/opensleep/opensleep/Services/Sonar/DiffSonarConsumer.swift
import Foundation

class DiffSonarConsumer {
    let sampleRate: Int
    private var iirFilter: IirFilter?
    private var prevRecord: [Float]?

    init(sampleRate: Int) {
        self.sampleRate = sampleRate
        let coefs = IirFilterDesignFisher.design(
            filterPassType: .highpass,
            filterCharacteristicsType: .butterworth,
            order: 18,
            ripple: -1.0,
            fcf1: 17000.0 / Double(sampleRate),
            fcf2: -1.0
        )
        self.iirFilter = IirFilter(coefficients: coefs)
    }

    struct ConsumerResult {
        let signal: Double
        let activity: Float
    }

    func processAndGetResult(_ fArr: [Float]) -> ConsumerResult {
        var filtered = [Float](repeating: 0.0, count: fArr.count)
        if let filter = iirFilter {
            for i in 0..<fArr.count {
                filtered[i] = Float(filter.step(Double(fArr[i])))
            }
        } else {
            filtered = fArr
        }

        var fSum: Float = 0.0
        if let fArr2 = prevRecord, fArr2.count == filtered.count {
            for i in 0..<filtered.count {
                fSum += abs(filtered[i] - fArr2[i])
            }
        }

        var fSum2: Float = 0.0
        for v in filtered {
            fSum2 += abs(v)
        }

        prevRecord = filtered
        return ConsumerResult(signal: Double(fSum2), activity: fSum)
    }

    func reset() {
        prevRecord = nil
        let coefs = IirFilterDesignFisher.design(
            filterPassType: .highpass,
            filterCharacteristicsType: .butterworth,
            order: 18,
            ripple: -1.0,
            fcf1: 17000.0 / Double(sampleRate),
            fcf2: -1.0
        )
        self.iirFilter = IirFilter(coefficients: coefs)
    }
}

File: android/app/src/main/java/app/opensleep/domain/sonar/DiffSonarConsumer.kt
package app.opensleep.domain.sonar

import app.opensleep.domain.dsp.IirFilter
import app.opensleep.domain.dsp.IirFilterDesignFisher
import app.opensleep.domain.dsp.FilterPassType
import app.opensleep.domain.dsp.FilterCharacteristicsType
import kotlin.math.abs

class DiffSonarConsumer(val sampleRate: Int) {
    private var iirFilter: IirFilter? = null
    private var prevRecord: FloatArray? = null

    init {
        val coefs = IirFilterDesignFisher.design(
            FilterPassType.highpass,
            FilterCharacteristicsType.butterworth,
            18,
            -1.0,
            17000.0 / sampleRate,
            -1.0
        )
        iirFilter = IirFilter(coefs)
    }

    data class ConsumerResult(val signal: Double, val activity: Float)

    fun processAndGetResult(fArr: FloatArray): ConsumerResult {
        val filtered = FloatArray(fArr.size)
        val filter = iirFilter
        if (filter != null) {
            for (i in fArr.indices) {
                filtered[i] = filter.step(fArr[i].toDouble()).toFloat()
            }
        } else {
            System.arraycopy(fArr, 0, filtered, 0, fArr.size)
        }

        val fArr2 = prevRecord
        val fSum = if (fArr2 != null && fArr2.size == filtered.size) {
            var s = 0.0f
            for (i in filtered.indices) {
                s += abs(filtered[i] - fArr2[i])
            }
            s
        } else {
            0.0f
        }

        var fSum2 = 0.0f
        for (v in filtered) {
            fSum2 += abs(v)
        }

        prevRecord = filtered.copyOf()
        return ConsumerResult(fSum2.toDouble(), fSum)
    }

    fun reset() {
        prevRecord = null
        val coefs = IirFilterDesignFisher.design(
            FilterPassType.highpass,
            FilterCharacteristicsType.butterworth,
            18,
            -1.0,
            17000.0 / sampleRate,
            -1.0
        )
        iirFilter = IirFilter(coefs)
    }
}
